crates/ruff/src/rules/flake8_return/rules.rs

```diff
@@ -354,10 +369,75 @@ fn unnecessary_assign(checker: &mut Checker, stack: &Stack, expr: &Expr) {
             return;
         }
 
-        checker.diagnostics.push(Diagnostic::new(
-            UnnecessaryAssign,
-            Range::from_located(expr),
-        ));
+        let mut diagnostic = Diagnostic::new(UnnecessaryAssign, Range::from_located(expr));
+
+        if checker.patch(diagnostic.kind.rule()) {
+            if let Some(assign_expr) = stack.assign_values.get(id.as_str()) {
+                if let Some(assign_locations) = stack.assigns.get(id.as_str()) {
+                    let expr_loc = expr.end_location.unwrap();
+                    if let Some(last_assign_loc) = assign_locations
+                        .iter()
+                        .rev()
+                        .find(|x| x.location < expr_loc)
+                    {
+                        let in_between_source = checker
+                            .locator
+                            .slice_source_code_range(&Range::new(
+                                last_assign_loc.end_location,
+                                expr.location,
+                            ))
+                            .to_owned();
+                        // Ensure that all indentation is identical (no if or while statement).
```

Comment:
I think we need a more rigorous way to detect this, if I'm interpreting the comment correctly. Indentation isn't a totally "safe" way to detect whether we have control flow like `if` or `while` etc., since users could do `if x: y` on a single line.

What's this logic guarding against? Maybe we can track this condition in some other way.

Reply:
Yeah, I see what you mean with the one-liner. The main problem is that there's no `Dedent` token during the code that ensures it's comments.

Reply:
I might have to be checking the line itself too (e.g. slicing the line and checking if there is any tokens there)